design/tls-client-verification.md

```diff
@@ -0,0 +1,161 @@
+# Client certificate validation (mTLS)
+
+Status: Draft
+
+## Goals
+
+- Allow client certificate validation (mTLS) on contour (performed by Envoy)
+- Allow various ways for client certificate validation; Spki, Hash and CA
+- Document mTLS configuration
+
+## Non Goals
+
+- Support for k8s ingress documents.
+- Repeat configuration details described in Envoy documentation (make references)
+
+## Background
+
+In TLS (and https) only the server is authenticated with a
+certificate, for instance you as a client can be sure that you speak
+with your bank and not some malicious site. But sometimes also the client
+must be authenticated. As noted in
+[wikipedia](https://en.wikipedia.org/wiki/Mutual_authentication)
+client certification (mTLS) is not very common for end-users but is
+more widespread for business-to-business (B2B) applications (which
+may use gRPC and REST APIs).
+
+I can't give a real-life example but it is easy to imagine cases where
+client validation is necessary, for instance for an admin interface to
+a server that is accesses by automated clients.
+
+Client certificate validation (mTLS) is supported by Envoy. It should
+be possible for `contour` users to utilize this feature.
+
+
+## High-Level Design
+
+At a high level I propose the following:
+
+1. A new record "clientValidation" is added in spec.virtualhost.tls
+   in the IngressRoute.
+
+2. The `clientValidation` contains configuration for client
+   certificate validation. Many ways of client certificate validation
+   may be specified (all ways supported by Envoy).
+
+
+### Sample YAML
+
+```
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata:
+  name: kahttp
+  namespace: default
+spec:
+  virtualhost:
+    fqdn: kahttp.com
+    tls:
+      secretName: contour-secret
+      clientValidation:
+        secretName: clientsecret
+        spkis:
+          - 2IEpPESU/mmC30tPsnOfbGKdwKdQfN/wZw1QWpjGlmk=
+        subjectAltNames:
+          - server.example.com
+  routes:
+    - match: /
+      services:
+        - name: kahttp
+          port: 80
+```
+
+
+## Detailed Design
+
+Since this design proposal required some "learning-by-doing" most of
+phase 1 is implemented by https://github.com/heptio/contour/pull/1226
+
+Unit-tests must be added on all appropriate places.
+
+
+### CAs in Secrets
+
+Same as for [TLS backend verification](tls-backend-verification.md).
```

Comment:
What is the same? How communication is secured? How secrets are named?

Reply:
They are named, used and created in the same way.

I removed the reference and copied the chapter from tls-backend-verification.md instead to make it clear.

Reply:
Communication is secured with TLS (encryption), same as always. mTLS is just a way to limit the access to the server to validated clients.